sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -73,109 +80,182 @@ public static BeamSqlEnv inMemory(TableProvider... tableProviders) {
+    ddl.execute(getContext());
   }
 
-  /** Register a UDF function which can be used in SQL expression. */
-  public void registerUdf(String functionName, Class<?> clazz, String method) {
-    connection.getCurrentSchemaPlus().add(functionName, UdfImpl.create(clazz, method));
+  public CalcitePrepare.Context getContext() {
+    return connection.createPrepareContext();
   }
 
-  /** Register a UDF function which can be used in SQL expression. */
-  public void registerUdf(String functionName, Class<? extends BeamSqlUdf> clazz) {
-    registerUdf(functionName, clazz, BeamSqlUdf.UDF_METHOD);
+  public Map<String, String> getPipelineOptions() {
+    return connection.getPipelineOptionsMap();
   }
 
-  /**
-   * Register {@link SerializableFunction} as a UDF function which can be used in SQL expression.
-   * Note, {@link SerializableFunction} must have a constructor without arguments.
-   */
-  public void registerUdf(String functionName, SerializableFunction sfn) {
-    registerUdf(functionName, sfn.getClass(), "apply");
+  public String explain(String sqlString) throws ParseException {
+    try {
+      return RelOptUtil.toString(planner.convertToBeamRel(sqlString));
+    } catch (Exception e) {
+      throw new ParseException("Unable to parse statement", e);
+    }
   }
 
-  /**
-   * Register a UDAF function which can be used in GROUP-BY expression. See {@link
-   * org.apache.beam.sdk.transforms.Combine.CombineFn} on how to implement a UDAF.
-   */
-  public void registerUdaf(String functionName, Combine.CombineFn combineFn) {
-    connection.getCurrentSchemaPlus().add(functionName, new UdafImpl(combineFn));
-  }
+  /** BeamSqlEnv's Builder. */
+  public static class BeamSqlEnvBuilder {
+    private String queryPlannerClassName =
+        "org.apache.beam.sdk.extensions.sql.impl.CalciteQueryPlanner";
 
-  /** Load all UDF/UDAF from {@link UdfUdafProvider}. */
-  public void loadUdfUdafFromProvider() {
-    ServiceLoader.<UdfUdafProvider>load(UdfUdafProvider.class)
-        .forEach(
-            ins -> {
-              ins.getBeamSqlUdfs().forEach((udfName, udfClass) -> registerUdf(udfName, udfClass));
-              ins.getSerializableFunctionUdfs()
-                  .forEach((udfName, udfFn) -> registerUdf(udfName, udfFn));
-              ins.getUdafs().forEach((udafName, udafFn) -> registerUdaf(udafName, udafFn));
-            });
-  }
+    private TableProvider initialTableProvider;
+    private String currentSchemaName;
+    private Map<String, TableProvider> schemaMap = new HashMap<>();
+    private Set<Map.Entry<String, Function>> functionSet = new HashSet<>();
 
-  public void loadBeamBuiltinFunctions() {
-    for (BeamBuiltinFunctionProvider provider :
-        ServiceLoader.load(BeamBuiltinFunctionProvider.class)) {
-      registerBuiltinUdf(provider.getBuiltinMethods());
+    public BeamSqlEnvBuilder setInitializeTableProvider(TableProvider tableProvider) {
```

Comment:
This is not optional, this should be a part of the builder constructor probably so that it's impossible to miss. And it's unclear why this is initial if there are no obvious ways to update it (the only other method to update it is called `addSchema` which is confusing). Rename to something like `setDefaultSchema()`?